add remaining diagnostics to the new diagnostics package

szy21 opened the pull request:
<!--- THESE LINES ARE COMMENTED -->
## Purpose 
<!--- One sentence to describe the purpose of this PR, refer to any linked issues:
#14 -- this will link to issue 14
Closes #2 -- this will automatically close issue 2 on PR merge
-->
Add EDMFX diagnostics and add some default diagnostics when running simulations with EDMFX. I think this includes all the important diagnostics we want to have.

## To-do
<!---  list of proposed tasks in the PR, move to "Content" on completion 
- Proposed task
-->
- [x] Check if the diagnostics are physically correct


## Content
<!---  specific tasks that are currently complete 
- Solution implemented
-->


<!---
Review checklist

I have:
- followed the codebase contribution guide: https://clima.github.io/ClimateMachine.jl/latest/Contributing/
- followed the style guide: https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/
- followed the documentation policy: https://github.com/CliMA/policies/wiki/Documentation-Policy
- checked that this PR does not duplicate an open PR.

In the Content, I have included 
- relevant unit tests, and integration tests, 
- appropriate docstrings on all functions, structs, and modules, and included relevant documentation.

-->

----
- [ ] I have read and checked the items on the review checklist.

Sbozzolo commented on config/model_configs/diagnostic_edmfx_bomex_box.yml:
```diff
@@ -26,4 +26,5 @@ z_max: 3e3
-dt_save_to_disk: "10mins"
\ No newline at end of file
+dt_save_to_disk: "10mins"
+output_default_diagnostics: true
```
Can you please add a newline?

Sbozzolo commented on src/diagnostics/default_diagnostics.jl:
```diff
@@ -92,7 +92,7 @@ daily_max(short_names; output_writer = HDF5Writer()) =
 """
 daily_mins(short_names...; output_writer = HDF5Writer()) =
-    common_diagnostics(24 * 60 * 60, min, output_writer, short_names...)
+    common_diagnostics(24 * 60 * 60, nothing, output_writer, short_names...)
```
I think this should be `min`. Why did you change to `nothing`?

szy21 replied:
Oh, this is a typo, thanks!

Sbozzolo commented on src/diagnostics/default_diagnostics.jl:
```diff
@@ -186,6 +186,24 @@ Return a `ScheduledDiagnostics` that computes the hourly average for the given v
+
+Return a list of `ScheduledDiagnostics` that compute the 10 min instantaneous output for the given variables.
+"""
+tenmin_insts(short_names...; output_writer = HDF5Writer()) =
+    common_diagnostics(10 * 60, nothing, output_writer, short_names...)
```
I think that this will return in some warnings being raised. With `common_diagnostics` you set `output_every`, but `compute_every` is set to `:timestep`. If you don't want time reductions, `compute_every` will be adjusted to be the same as `output_every`, but a warning will be produced.

So, I think that you have to change this to
```suggestion
function tenmin_insts(short_names...; output_writer = HDF5Writer()) 
period = 10 * 60 # seconds
        return [
        ScheduledDiagnosticTime(
            variable = get_diagnostic_variable(short_name),
            compute_every = period,
            output_every = period, 
            output_writer = output_writer,
        ) for short_name in short_names
    ]

end
```

szy21 replied:
Thanks!

Sbozzolo reviewed (commented):
Looks good!

My main comment (other than the tiny details below) is that you the "comment" field in your variables is not really useful since you just paraphrase the `long_name`.

Maybe we can copy and paste the comments from the airtable with all the CMIP quantities?

szy21 commented on src/diagnostics/core_diagnostics.jl:
```diff
@@ -261,6 +261,70 @@ add_diagnostic_variable!(
+add_diagnostic_variable!(
+    short_name = "cli",
+    long_name = "Mass Fraction of Cloud Ice",
+    standard_name = "mass_fraction_of_cloud_icein_air",
```
```suggestion
    standard_name = "mass_fraction_of_cloud_ice_in_air",
```

szy21 commented:
> Maybe we can copy and paste the comments from the airtable with all the CMIP quantities?

All the EDMFX specific diagnostics don't have a corresponding variable in the CMIP table, as they don't use the same convection scheme in CMIP. The comments for all updraft variables (diagnostics ending with `up`) are mostly to just describe that the diagnostics is for the first updraft. The comments for environment variables are basically the `long_name`, maybe I can just remove them?

Sbozzolo commented:
> > Maybe we can copy and paste the comments from the airtable with all the CMIP quantities?
> 
> All the EDMFX specific diagnostics don't have a corresponding variable in the CMIP table, as they don't use the same convection scheme in CMIP. The comments for all updraft variables (diagnostics ending with `up`) are mostly to just describe that the diagnostics is for the first updraft. The comments for environment variables are basically the `long_name`, maybe I can just remove them?

When I look the [table](https://airtable.com/appYNLuWqAgzLbhSq/shrKcLEdssxb8Yvcp/tblL7dJkC3vl5zQLb), I see for example for `cli`:
```
Includes both large-scale and convective cloud. Calculate as the mass of cloud liquid water in the grid cell divided by the mass of air (including the water in all phases) in the grid cells. Precipitating hydrometeors are included ONLY if the precipitating hydrometeors affect the calculation of radiative transfer in model.
```

I don't know how much detail we should have, but I think that having more is better than having less (especially if the variable is never going to change).

szy21 commented:
> 
> When I look the [table](https://airtable.com/appYNLuWqAgzLbhSq/shrKcLEdssxb8Yvcp/tblL7dJkC3vl5zQLb), I see for example for `cli`:
> 
> ```
> Includes both large-scale and convective cloud. Calculate as the mass of cloud liquid water in the grid cell divided by the mass of air (including the water in all phases) in the grid cells. Precipitating hydrometeors are included ONLY if the precipitating hydrometeors affect the calculation of radiative transfer in model.
> ```
> 
> I don't know how much detail we should have, but I think that having more is better than having less (especially if the variable is never going to change).

Oh, right, I thought you meant the EDMFX variables. Yes, I can add more comments to variables like `clw` and `cli`. One reason I didn't add it is that the comments from the table don't apply to our model yet. For example, for `cli`, we don't have convective cloud, or precipitating hydrometeors. But maybe we can just add the comments and change how the diagnostics are calculated later?

Sbozzolo commented:
> > When I look the [table](https://airtable.com/appYNLuWqAgzLbhSq/shrKcLEdssxb8Yvcp/tblL7dJkC3vl5zQLb), I see for example for `cli`:
> > ```
> > Includes both large-scale and convective cloud. Calculate as the mass of cloud liquid water in the grid cell divided by the mass of air (including the water in all phases) in the grid cells. Precipitating hydrometeors are included ONLY if the precipitating hydrometeors affect the calculation of radiative transfer in model.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I don't know how much detail we should have, but I think that having more is better than having less (especially if the variable is never going to change).
> 
> Oh, right, I thought you meant the EDMFX variables. Yes, I can add more comments to variables like `clw` and `cli`. One reason I didn't add it is that the comments from the table don't apply to our model yet. For example, for `cli`, we don't have convective cloud, or precipitating hydrometeors. But maybe we can just add the comments and change how the diagnostics are calculated later?

I think we should describe what we have, and update the description when we update the diagnostic.

szy21 commented:
@Sbozzolo I updated the comment of several variables following the CMIP air table, and removed the comment of some variables where they are similar to the long name. I also set the default EDMF diagnostics to daily average, and included an example yaml file in ci that calculates 10-min instantaneous EDMF diagnostics. The results look physical. Would you like to take another look?

Sbozzolo commented:
> @Sbozzolo I updated the comment of several variables following the CMIP air table, and removed the comment of some variables where they are similar to the long name. I also set the default EDMF diagnostics to daily average, and included an example yaml file in ci that calculates 10-min instantaneous EDMF diagnostics. The results look physical. Would you like to take another look?

It looks good to me. The only thing: could you split the extremely long strings to multiline strings?

Also, I didn't realize how many EDMF examples we are running. In the future, if we want the whole suite of diagnostics for all the EMDF examples, we will probably have to define them as default in the code. (But this is good for now).

szy21 commented:
> It looks good to me. The only thing: could you split the extremely long strings to multiline strings?

I changed some of the comment to multiline strings. Is that what you mean?

> Also, I didn't realize how many EDMF examples we are running. In the future, if we want the whole suite of diagnostics for all the EMDF examples, we will probably have to define them as default in the code. (But this is good for now).

Yeah, I think it would be good if we don't need to specify them for each job. I wonder if the config can take multiple yaml files now (it could before I think). If so, we can have one common diagnostic yaml file for the current edmf examples.

szy21 commented:
bors r+

szy21 commented:
bors r+

simonbyrne commented:
Part of #2200
